File: spark_app/process_data.py
import os
from pyspark.sql import SparkSession

def main():
    print("Start PySpark Job for E-commerce Transactions file...")

    # [CHECK] each argument
    # Initialize SparkSession with Iceberg and MinIO (S3) configuration
    spark = SparkSession.builder \
    .appName("Iceberg-MinIO-Transactions-Job") \
    .config("spark.sql.extentions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions") \
    .config("spark.sql.catalog.local", "org.apache.iceberg.spark.SparkCatalog") \
    .config("spark.sql.catalog.local.type", "hadoop") \
    .config("spark.sql.catalog.local.warehouse", "s3a://lakehouse-bucket/warehouse") \
    .config("spark.hadoop.fs.defaultFS", "file:///") \
    .config("spark.hadoop.fs.s3a.endpoint", "http://my-minio:9000") \
    .config("spark.hadoop.fs.s3a.access.key", "admin") \
    .config("spark.hadoop.fs.s3a.secret.key", "admin123") \
    .config("spark.hadoop.fs.s3a.path.style.access", "true") \
    .config("spark.hadoop.fs.s3a.connection.ssl.enabled", "false") \
    .config("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem") \
    .getOrCreate()
    
    print("Spark Session Initialized Successfully")

    # 1. Read the local CSV file from the container's filesystem
    csv_path = "file:///app/transactions_data.csv"
    print(f"Reading data from {csv_path}")
    
    df = spark.read.option("header", "true") \
        .option("inferSchema", "true") \
        .csv(csv_path)
    
    print(f"Loaded {df.count()} rows successfully. Schema:")
    df.printSchema()
    
    # 2. Write data into Apache Iceberg format in MinIO
    table_name = "local.db.transactions"
    print(f"Writing data to Iceberg table: {table_name}...")

    # [CHECK] not needed
    spark.sql("CREATE NAMESPACE IF NOT EXISTS local.db")
    # Using append mode for run the DAG multiple times and see data grows
    df.write.format("iceberg") \
        .mode("append") \
        .saveAsTable(table_name)

    print("Data successfully written into Iceberg via MinIO!")

    # 3. Read back to verify
    print("Reading top 5 rows back from Iceberg...")
    read_df = spark.read.format("iceberg").load(table_name)
    read_df.show(5)

    spark.stop()

if __name__ == "__main__":
    main()

File: spark_scala/spark_scala/src/main/scala/com/lakehouse/ProcessData.scala
package com.lakehouse

import org.apache.spark.sql.SparkSession

// create a SparkSession singleton object
object ProcessData {
    def main(args: Array[String]): Unit ={
        println("Starting Spark Session")

        // inisitialize SparkSession
        val spark = SparkSession.builder()
            .appName("Scala-Iceberg-MinIO-Transaction-Job")
            .config("spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions")
            .config("spark.sql.catalog.local", "org.apache.iceberg.spark.SparkCatalog")
            .config("spark.sql.catalog.local.type", "hadoop")
            .config("spark.sql.catalog.local.warehouse", "s3a://lakehouse-warehouse")
            .config("spark.hadoop.fs.defaultFS", "file:///")
            .config("spark.hadoop.fs.s3a.endpoint", "http://my-minio:9000")
            .config("spark.hadoop.fs.s3a.access.key", "admin")
            .config("spark.hadoop.fs.s3a.secret.key", "admin123")
            .config("spark.hadoop.fs.s3a.path.style.access", "true")
            .config("spark.hadoop.fs.s3a.connection.ssl.enabled", "false")
            .config("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
            .getOrCreate()

        println("Spark Session Started")

        // 0. define Schema for the CSV file
        val mySchema = StructType(Array(
            StructField("transaction_id", StringType, nullable = false), // is nullable
            StructField("user_id", IntegerType, nullable = false),
            StructField("timestamp", TimestampType, nullable = false),
            StructField("product_category", StringType, nullable = false),
            StructField("amount", DoubleType, nullable = false),
            StructField("payment_method", StringType, nullable = false),
            StructField("status", StringType, nullable = false)
        ))


        // 1. read local CSV file into a DataFrame
        val csvPath = "file:///app/transactions_data.csv"
        println(s"Reading CSV file from path: $csvPath")

        val df = spark.read
            .option("header", "true")
            .option("timestampFormat", "yyyy-MM-dd HH:mm:ss")
            .option("mode", "FAILFAST")
            .schema(mySchema)
            .csv(csvPath)

        println(s"Loaded ${df.count()} rows from CSV file into DataFrame. Schema:")
        df.printSchema() // Eager

        // 2. write the DataFrame to Iceberg table in MinIO
        val tableName = "local.db.transactions"
        println(s"Writing DataFrame to Iceberg table: $tableName")

        spark.sql("CREATE NAMESPACE IF NOT EXISTS local.db")

        df.write
            .format("iceberg")
            .mode("overwrite")
            .saveAsTable(tableName)

        println(s"DataFrame written to Iceberg table: $tableName")

        // 3. read & validate the data from Iceberg table
        println(s"Reading data back from Iceberg table: $tableName")
        val readDf = spark.read.format("iceberg").load(tableName)
        readDf.show(5)

        spark.stop()
    }
}